docs: write error throwing and handling guides (#31)
Financial-Times/dotcom-reliability-kit: docs/getting-started/handling-errors.md

# Getting started: handling errors

We've covered creating and throwing good errors, now let's talk about handling errors. The way you handle errors, particularly in Express, is important. You can end up writing a lot of code in different parts of your app if you're not careful. More code isn't always bad, but duplicated error handling can lead to your errors not being surfaced to where they're needed.

  * [Bubbling up in Express](#bubbling-up-in-express)
  * [Don't fear the try/catch block](#dont-fear-the-trycatch-block)
    * [Error checking](#error-checking)
    * [Multiple try/catch blocks](#multiple-trycatch-blocks)
  * [Handling recoverable errors](#handling-recoverable-errors)

## Bubbling up in Express

In Express, error handlers can be bound at the Application and Route level. This allows you to write lots of error handling code in one place, and you can rely on your Express error handler to do things such as [log the appropriate error detail](./logging-errors.md), forward them on to a service like Splunk, or render a custom error page.

> The following is not a great pattern, as Express has no chance to log the error to a central place, we're missing a lot of information that we should be capturing:

Reviewer: **nitpick**: comma splice, "we're missing..." etc should be a new sentence

```suggestion
The following is not a great pattern, as Express has no chance to log the error to a central place. We're missing a lot of information that we should be capturing:
```

Reviewer: now i think about it more, maybe the sentence should be restructured, because "We're missing a lot of information that we should be capturing" shouldn't be the thing that's introducting the example via the colon

Reviewer 2: desperate for a ☕ here

Reviewer: your face is desparate for a coffee

Author: @apaleslimghost how about this?

```suggestion
The following is not a great pattern, as Express has no chance to log the error to a central place. This means our logs could be missing information that's essential to debugging this error:
```

```js
app.get('/', (request, response) => {
    try {
        exampleFunctionWhichThrowsAnError();
    } catch (error) {
        response.status(500).send(error.message);
    }
});
```

The way you make sure errors are handled by the main Express error handler is by always passing errors on. The way you do this will depend on whether your route handler is `async` or not.

For non-`async` route handlers, any errors thrown will immediately be handled by Express:

```js
// This is better because Express will get hold of the error, potentially
// log it, and render a page with a `500` status code
app.get('/', (request, response) => {
    exampleFunctionWhichThrowsAnError();
});
```

> For `async` route handlers, it's very important that you either use a module like [async-express-errors](https://www.npmjs.com/package/express-async-errors) (to avoid app crashes) or wrap your handler in an outer try/catch block:

Author: I see where you're coming from but I think because we can't _mandate_ that teams user `express-async-errors` I'd like to keep examples assuming that they don't. I think it might be worth adding an entirely new documentation page outside of the "getting started" guide which covers `async` route handlers and then linking to it from here.

I think I can add in a few clarifying messages to explain to the reader that we're making the assumption that they're not monkey-patching Express, and also that if they _are_ then they can ignore the `next` pattern.

> For `async` route handlers, it's very important that you either use a module like [async-express-errors](https://www.npmjs.com/package/express-async-errors) (to avoid app crashes) or wrap your handler in an outer try/catch block:

Author: I've updated the docs to better explain this 👍  thanks. See the most recent commit! Feel free to unresolve if you think we need more

```js
app.get('/', async (request, response, next) => {
    try {
        await exampleFunctionWhichThrowsAnError();
    } catch (error) {
        // With a wrapping try/catch block, it's important to
        // pass the error onto the next route handler so that
        // it can be logged and an appropriate error page can
        // be rendered
        next(error);
    }
});
```

Reliability Kit provides some Express error handlers which can be used to [log errors in a consistent and centralised way](./logging-errors.md).

> 

Author: I do think we should document this stuff for sure, but maybe not in the getting started guide? The more I think about the more I want to add, maybe we need a separate FAQ-like document for specific cases? Or maybe it's OK for teams to make their own call for now and we can document the patterns that come out of this?

## Don't fear the try/catch block

> One of the tools we have to make sure [thrown errors should be specific](./throwing-errors.md#being-specific) is the `try`/`catch` block. In previous examples we've wrapped all our code in a single try/catch block and sent the error on to the Express error handler. If we want to be _really_ specific about what went wrong in our code, then this is often not enough.

Reviewer: **nitpick**: consistency in code quotes

```suggestion
One of the tools we have to make sure [thrown errors should be specific](./throwing-errors.md#being-specific) is the `try`/`catch` block. In previous examples we've wrapped all our code in a single `try`/`catch` block and sent the error on to the Express error handler. If we want to be _really_ specific about what went wrong in our code, then this is often not enough.
```

        // Make an API request using an example API client
        const fruit = await fruitApiClient.getByName(request.params.name);
        response.send(fruit);

    } catch (error) {
        next(error);
    }
});
```

In this scenario we have the following possible errors which we know can make it into our final try/catch:

  * The `name` parameter is an empty string, throwing an `AssertionError`

  * The fruit API is down and the client throws an error with a `statusCode` property set to `503`

  * The fruit API does not have a fruit matching `request.params.name` and so the client throws an error with a `statusCode` property set to `404`

Let's update our `catch` block to handle these cases (Note: this example uses `HttpError` from the [`@dotcom-reliability-kit/errors`](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#readme) package):

```js
// E.g. GET https://your-app/fruit/feijoa
app.get('/fruit/:name', async (request, response, next) => {
    try {
        // [The code in the example above, excluded for brevity]
    } catch (error) {
        
        // If the name parameter is not set, let's throw a more

>         // friendly 404 error. We know that we're dealing with

Reviewer: **nitpick**: should say `400 error`

```suggestion
        // friendly 400 error. We know that we're dealing with
```

        // with some added detail
        if (error.statusCode === 404) {
            return next(new HttpError({
                statusCode: 404,
                code: 'FRUIT_NOT_FOUND',
                message: `The fruit ${request.params.name} was not found in the fruit API`
            }));
        }

        // If the fruit API was down, let's throw an error which
        // better describes what's happened. We switch the status
        // code to 502 because that matches what's happening – we
        // received an invalid response from an upstream server.
        // We can also indicate the system which caused the error
        if (error.statusCode === 503) {
            return next(new HttpError({
                statusCode: 502,
                code: 'FRUIT_API_FAILED',
                message: 'The fruit API was not reachable',
                relatesToSystems: ['fruit-api']
            }));
        }

        // We default to passing on the error as-is, because
        // this is probably a programmer error rather than
        // operational and we want to let the app decide how
        // to handle it later
        next(error);
    }
});
```

> With this checking in place you can be confident that your error logging (and any dashboards powered by it) will display easy to understand errors and make debugging much easier. However this checking is quite simplistic. What if we make multiple API requests, any of which can fail?

Reviewer: **nitpick**: structure

```suggestion
With this checking in place you can be confident that your error logging (and any dashboards powered by it) will display easy to understand errors and make debugging much easier. However, this checking is quite simplistic; what if we make multiple API requests, any of which can fail?
```

                    message: 'The fruit API was not reachable',
                    relatesToSystems: ['fruit-api']
                });
            }
            // We still want to throw if we have a different error
            // from this service
            throw error;
        }

    } catch (error) {

        // Handle the 404 case
        if (error.statusCode === 404) {
            return next(new HttpError({
                statusCode: 404,
                code: 'FAVOURITE_FRUIT_NOT_FOUND',
                message: `A favourite fruit cannot be found for ${request.params.person}`
            }));
        }

        // If we get here, we could have either an unknown programmer error or
        // one of our specially thrown 502 errors
        next(error);
    }
});
```

With this checking and multiple `try`/`catch` blocks in place you can be confident about exactly which part of your app is failing. If an API is down then you get a message which specifically outlines which. Your tooling/reporting can also confidently say X% of requests fail because this specific API is down a lot and you can prioritise some work to address this.

## Handling recoverable errors

> Sometimes an error in your code is recoverable, and a page can continue to be rendered even if something failed. We call these recoverable errors. Even when an error is recoverable, it's still useful to have logs to help us understand how frequently certain parts of a website are missing.

Author: Yeah definitely, once I'm done with error logging I want to make sure we provide _something_ more concrete for logging "grey failure"/recoverable errors, this doc will be edited further once I PR new packages in to make sure it stays up to date with what we have codified.
